optimum/intel/openvino/modeling_decoder.py

```diff
@@ -463,32 +461,41 @@ def forward(
         attention_mask: Optional[torch.LongTensor] = None,
         past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
         position_ids: Optional[torch.LongTensor] = None,
+        infer_context: Optional[List[openvino.runtime.InferRequest]] = None,
         **kwargs,
     ) -> CausalLMOutputWithPast:
-        self.compile()
-
         inputs = self.prepare_inputs(
             input_ids=input_ids,
             attention_mask=attention_mask,
             past_key_values=past_key_values,
             position_ids=position_ids,
             **kwargs,
         )
-
         # Run inference
-        self.request.start_async(inputs, share_inputs=True)
-        self.request.wait()
-        logits = torch.from_numpy(self.request.get_tensor("logits").data).to(self.device)
+        if self.stateful and past_key_values is not None:
+            # for stateful models, infer request is created in generate and __call_ methods and passed in the cycle via past_key_values param
+            infer_request = past_key_values[1]
+        else:
+            if infer_context is not None:
+                infer_request = infer_context[
+                    0
+                ]  # Use passed inference request if provided in kwargs, create new one overwise
+            else:
+                self.compile()
+                infer_request = self.compiled_model.create_infer_request()
+        infer_request.start_async(inputs, share_inputs=True)
+        infer_request.wait()
+        logits = torch.from_numpy(infer_request.get_tensor("logits").data).to(self.device)
         if self.stateful:
             # Need a marker to differentiate the first generate iteration from the others in
             # the first condition at the function beginning above.
             # It should be something that is not None and it should be True when converted to Boolean.
-            past_key_values = ((),)
+            past_key_values = ((inputs["beam_idx"]), infer_request)
```

Comment:
not related to `past_key_values` so I don't think we should update `past_key_values` here, the resulting output will not be what it's expected for example :
```
output = model(**tokens)
pkv = output.past_key_values
```

Reply:
My point is that it's not related to `past_key_values` so we shouldn't update it with beam_idx / inference execution context

Reply:
@slyalin can you add your comments here? The idea was to reused this variable for stateful models because they don't use it at all. That was the only method we found that could be used to pass the beam_idx and execution context (which includes the state data) without changing the model API. The other alternative was with using model.clone() method for each thread which would also using a separate execution context without duplicating memory consumption https://github.com/huggingface/optimum-intel/pull/564. Would cloning be better method to support concurrency in the execution? Is there some other option we are not aware of? I guess it is a bit unique situation with the stateful models in openvino so probably it is not handled in transformers lib.

Reply:
> My point is that it's not related to past_key_values

Definitely it _is_ related to `past_key_values` even more than the old `((),)` value. `beam_idx` together with `infer_request` are used to track `past_key_values` for a particular sequence. Literally, `infer_request` has a model state that consists of `past_key_values` tensors, and `beam_idx` allows indirect rows reordering in that state in case of beam search. This PR just makes it more explicit than it was before and moving this attributes from the model class instance to each sequence, which allows having multiple sequences for a single model class instance.

@echarlaix, do you have a better alternative to pass these values?

Reply:
> we shouldn't update this variable with beam_idx / inference execution context

I saw another example where `beam_idx` is stored directly in `past_key_values` here: https://github.com/huggingface/optimum-intel/pull/567/files#diff-96778abd26ec589f358d62af391a34fdda4f7e4734cca48b421c567f5f3042e6R591-R600. So indeed it was treated as a detail of implementation that depends on backend used for inference.

Reply:
> @echarlaix, do you have a better alternative to pass these values?

why not introducing a class inheriting from `ModelOutput` like `CausalLMOutputWithPast` https://github.com/huggingface/transformers/blob/536ea2aca234fb48c5c69769431d643b0d93b233/src/transformers/modeling_outputs.py#L678 with dedicated `beam_idx` / `inference_request` arguments ?

Reply:
@echarlaix, thanks for the suggestion. But not sure that I understand the flow in this case. Do you have examples where such an inheritance is used for similar purposes?

Reply:
here for example : [BertForPreTrainingOutput](https://github.com/huggingface/transformers/blob/536ea2aca234fb48c5c69769431d643b0d93b233/src/transformers/models/bert/modeling_bert.py#L738)

Reply:
@echarlaix if we create new modelOutput data class and it is returned by the Forward method, how it could be passed back to the Forward method in the next cycle?

Reply:
can you try something like :
```python
from dataclasses import dataclass
from transformers.modeling_outputs import ModelOutput

@dataclass
class CausalLMOutputWithPast(ModelOutput):
    logits: torch.FloatTensor = None
    past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None
    beam_idx: Optional[int] = None
    inference_session = None
```



and then overwritte `_update_model_kwargs_for_generation` https://github.com/huggingface/transformers/blob/45c065109074d60c587d3e562f16531d02a422f6/src/transformers/generation/utils.py#L630 by adding somethign like :   

```python
    def _update_model_kwargs_for_generation(
        self,
        outputs: ModelOutput,
        model_kwargs: dict[str],
        is_encoder_decoder: bool = False,
        standardize_cache_format: bool = False,
    ) -> dict[str]:
        model_kwargs = super()._update_model_kwargs_for_generation(
            outputs=outputs,
            model_kwargs=model_kwargs,
            is_encoder_decoder=is_encoder_decoder,
            standardize_cache_format=standardize_cache_format,
        )

        if "beam_idx" in outputs:
            model_kwargs["beam_idx"] = outputs["beam_idx"]

        return model_kwargs
```

Reply:
Let me know if you need help on this @dtrawins

Reply:
@echarlaix @eaidova @slyalin Could you have a look if the latest version is passing the context fine now?
I'm not reusing past_key_values for stateful models with th generation context. There are additional fields in the `forward` output beam_idx and infer_request. Now only 9 tests is left to fix but seams unrelated to concurrency. Probably rebase from main is needed.
Anyway can one comment if beam_idx would be populated correctly. It is not defined now in reorder_caches for stateful models.

Reply:
What i confirmed is beam_idx was not passed correctly. The same, initial beam_idx was circulating for the whole pipeline resulting in incorrect accuracy with beam search. Somehow it was not detected by functional tests. 
Anyway my proposal is to pass the beam_idx content from reorder_caches method inside past_key_value. I tested it gives correct results and the code is in my opinion clean. The forward method returns empty past_key_values as expected for stateful models. In case someone would like to manage the pipeline for stateless models outside of transformers using just forward method, it would be still possible - beam_idx should be passed inside past_key_value and inference_request context via model_args. Anyway that is probably unlikely use case scenario. Would it be acceptable?